src/keystore.c

```diff
@@ -43,54 +108,79 @@ struct _DPS_MemoryKeyStore {
     size_t entriesCount;
     size_t entriesCap;
 
+    uint8_t* networkId;
+    size_t networkIdLen;
     uint8_t* networkKey;
     size_t networkKeyLen;
+
+    unsigned char *ca;
+    size_t caLen;
+
+    unsigned char *cert;
+    size_t certLen;
+    unsigned char *privateKey;
+    size_t privateKeyLen;
+    unsigned char *password;
+    size_t passwordLen;
 };
 
-static DPS_Status MemoryKeyStoreNetworkKeyCallback(DPS_KeyStore* keyStore, uint8_t* buffer, size_t bufferLen, size_t* keyLen)
+static DPS_Status MemoryKeyStoreKeyAndIdentityHandler(DPS_KeyStoreRequest* request)
 {
-    DPS_MemoryKeyStore* mks = keyStore->userData;
-    if (!mks->networkKey) {
+    DPS_MemoryKeyStore* mks = (DPS_MemoryKeyStore*)DPS_KeyStoreHandle(request);
+    if (!mks->networkId || !mks->networkKey) {
         return DPS_ERR_MISSING;
     }
-
-    if (bufferLen < mks->networkKeyLen) {
-        return DPS_ERR_RESOURCES;
-    }
-
-    memcpy(buffer, mks->networkKey, mks->networkKeyLen);
-    *keyLen = mks->networkKeyLen;
-
-    return DPS_OK;
+    return DPS_SetKeyAndIdentity(request, mks->networkKey, mks->networkKeyLen,
+                                 mks->networkId, mks->networkIdLen);
 }
 
-static DPS_Status MemoryKeyStoreContentKeyCallback(DPS_KeyStore* keyStore, const DPS_UUID* kid, uint8_t* key, size_t keyLen)
+static DPS_Status MemoryKeyStoreKeyHandler(DPS_KeyStoreRequest* request, const unsigned char* id, size_t len)
 {
-    DPS_MemoryKeyStore* mks = keyStore->userData;
+    DPS_MemoryKeyStore* mks = (DPS_MemoryKeyStore*)DPS_KeyStoreHandle(request);
     size_t i;
-    for (i = 0; i < mks->entriesCount; ++i) {
-        DPS_MemoryKeyStoreEntry* entry = mks->entries + i;
-        if (DPS_UUIDCompare(kid, &entry->kid) == 0) {
-            if (keyLen < entry->keyLen) {
-                DPS_ERRPRINT("Key has size %d, but callback given only %d buffer", entry->keyLen, keyLen);
-                return DPS_ERR_MISSING;
+
+    if (len == sizeof(DPS_UUID)) {
+        for (i = 0; i < mks->entriesCount; ++i) {
+            DPS_MemoryKeyStoreEntry* entry = mks->entries + i;
```

Comment:
could this be accessing one past the end of the mks->entries?

Reply:
i < mks->entriesCount so I think this is safe.